### ISSUE TITLE
Add masking support for ParentIndependentBinarySelectionModel to exclude specific sites from selection factor calculations

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/netam#165]
## Background
The `ParentIndependentBinarySelectionModel` class currently receives a `mask` parameter in its forward method but does not use it at all. The model uses `amino_acid_indices` only to determine batch size and sequence length, while the mask (which indicates valid amino acid sites) is completely ignored. This prevents the model from properly handling masked sites during training and inference, which is essential for:
- Excluding padding tokens from loss calculations
- Handling variable-length sequences
- Ignoring specific positions (e.g., conserved sites, framework regions)

## Current Behavior
```python
def forward(self, amino_acid_indices: Tensor, mask: Tensor) -> Tensor:
    batch_size, seq_len = amino_acid_indices.shape  # Uses amino_acid_indices for shape
    # ... rest of implementation
    # The mask parameter is never used!
```

### 1. Modify the forward method to apply masking
```python
def forward(self, amino_acid_indices: Tensor, mask: Tensor) -> Tensor:
    batch_size, seq_len = amino_acid_indices.shape
    position_factors = self.log_selection_factors[:seq_len]
    
    if self.output_dim == 1:
        result = position_factors.expand(batch_size, seq_len)
        # Apply masking: multiplicative masking in log space (consistent with other models)
        result = result * mask
    else:
        result = position_factors.unsqueeze(0).expand(batch_size, seq_len, self.output_dim).clone()
        
        # Apply wildtype zapping if applicable
        if self.output_dim >= 20 and self.wildtype_aa_idxs is not None:
            wt_idxs_batch = self.wildtype_aa_idxs[:seq_len].unsqueeze(0).expand(batch_size, -1)
            result = zap_predictions_along_diagonal(result, wt_idxs_batch, fill=0.0)
        
        # Apply masking: expand mask to match output dimensions
        result = result * mask.unsqueeze(-1)
    
    return result
```

### 2. Consider gradient handling during training
- Masked positions (where mask is False) should not contribute to gradients
- Setting masked positions to 0 in log space will result in 0 after exponentiation
- This naturally prevents gradient flow through masked positions

## Testing Requirements

1. **Unit tests for mask application**:
   - Verify that positions where mask=False have the mask_value
   - Verify that positions where mask=True retain their selection factors
   - Check gradient flow is blocked for masked positions
   - Test both 1D and multi-dimensional output cases

2. **Integration tests**:
   - Ensure compatibility with existing training pipelines
   - Verify loss calculation excludes masked positions
   - Test variable-length sequence handling

3. **Edge cases**:
   - All positions masked (mask all False)
   - No positions masked (mask all True)
   - Partial masking patterns

## Compatibility Considerations

1. **Backward compatibility**: 
   - Default behavior when mask is all True should match current behavior
   - Existing saved models should load correctly

2. **Consistency with other models**:
   - Other models in the codebase apply masking (e.g., line 126 in `FivemerModel`: `rates = torch.exp(log_kmer_rates * masks)`)
   - Note: Some models multiply by mask, others may use different approaches
   - Ensure similar masking semantics across model classes

## Implementation Steps

1. [ ] Implement mask application in forward method
2. [ ] Add configuration options for mask handling
3. [ ] Write comprehensive unit tests
4. [ ] Test integration with existing training code
5. [ ] Update documentation and docstrings
6. [ ] Ensure backward compatibility with saved models

## Alternative Approaches Considered

1. **Multiplicative masking in log space**: Multiply log selection factors by mask
   - Pros: Similar to other models in codebase
   - Cons: 0 * log_factor = 0, and exp(0) = 1, which may not be desired

2. **Multiplicative masking in linear space**: Apply mask after exponentiation
   - Pros: Clear semantics (masked positions get 0 selection)
   - Cons: Requires changes outside forward method

3. **Separate masked parameter tensor**: Learn different parameters for maskable positions
   - Pros: More flexible
   - Cons: Increases model complexity, requires retraining

## Questions for Discussion

1. Should the mask value be configurable or fixed to `-inf`?
2. How should masking interact with the wildtype sequence zapping?
3. Should we follow the multiplicative pattern used by other models (e.g., `FivemerModel`) or use the `-inf` approach?
4. Do we need to handle the mask differently during training vs inference?

## References
- Other models using masks: 
  - `FivemerModel` (line 126): `rates = torch.exp(log_kmer_rates * masks)`
  - `RSFivemerModel` (line 147): `rates = torch.exp(log_kmer_rates * masks)`
  - `SHMoofModel` (line 179): `rates = torch.exp((log_kmer_rates + log_site_rates) * masks)`
- Mask tensor creation: `aa_mask_tensor_of` function in sequences module
- Current `ParentIndependentBinarySelectionModel` implementation: lines 1010-1095
